src/drawers/lineDrawer.ts

```diff
@@ -75,4 +90,103 @@ export class Line extends Drawer {
 
     context.restore();
   }
+
+  /**
+   * Collapse line geometry
+   *
+   * Assuming that there are many points that are drawn on the same coordinate,
+   * we can save a lot of render time by just drawing one line from the min to
+   * max y-coordinate of all those points.
+   */
+  private _collapseLineGeometry(ctx: CanvasRenderingContext2D, data: any[]) {
+    const d3Line = this._d3LineFactory();
+    const xFn = d3Line.x();
+    const yFn = d3Line.y();
+
+    // move to valid start
+    let i = 0;
+    for (; i <= data.length; ++i) {
+      const d = data[i];
+      if (d != null) {
```

Comment:
use the function returned by `d3Line.defined()` to determine whether the data point is valid or not